hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java

```diff
@@ -438,33 +424,48 @@ private void writeTableMetadataForTableServices(HoodieTable<T, JavaRDD<HoodieRec
     UpgradeDowngrade upgradeDowngrade = new UpgradeDowngrade(
         metaClient, config, context, SparkUpgradeDowngradeHelper.getInstance());
     if (upgradeDowngrade.needsUpgradeOrDowngrade(HoodieTableVersion.current())) {
-      if (config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()) {
+      try {
+        // Lock the upgrade step and the follow-on metadata table creation
+        // and the initial bootstrapping so that concurrent writers if any
+        // are blocked from racing with these one time operations.
         this.txnManager.beginTransaction();
-        try {
-          // Ensure no inflight commits by setting EAGER policy and explicitly cleaning all failed commits
-          this.rollbackFailedWrites(getInstantsToRollback(metaClient, HoodieFailedWritesCleaningPolicy.EAGER, Option.of(instantTime)), true);
-          new UpgradeDowngrade(
-              metaClient, config, context, SparkUpgradeDowngradeHelper.getInstance())
+        if (config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()) {
+          // Ensure no in-flight commits by setting EAGER policy and explicitly cleaning all failed commits
+          this.rollbackFailedWrites(getInstantsToRollback(metaClient,
+              HoodieFailedWritesCleaningPolicy.EAGER, Option.of(instantTime)), true);
+          new UpgradeDowngrade(metaClient, config, context, SparkUpgradeDowngradeHelper.getInstance())
               .run(HoodieTableVersion.current(), instantTime);
-        } finally {
-          this.txnManager.endTransaction();
+        } else {
+          upgradeDowngrade.run(HoodieTableVersion.current(), instantTime);
         }
-      } else {
-        upgradeDowngrade.run(HoodieTableVersion.current(), instantTime);
+        metaClient.reloadActiveTimeline();
+        initializeMetadataTable(Option.of(instantTime));
```

Comment:
this already takes care of my suggestion on the other one.

Reply:
right.